### ISSUE TITLE
file descriptor leak

### ISSUE THREAD

**Issue by vtjnash** [JuliaWeb/HTTP.jl#26]
I recommend setting a low ulimit first, so that this doesn't take a long time (`ulimit -n 60`)
```julia
julia> for i in 1:10000
         HTTP.get("google.com")
       end
ERROR (unhandled task failure): DNSError: google.com, unknown node or service (EAI_NONAME)
Stacktrace:
 [1] getaddrinfo(::String) at socket.jl:649
 [2] (::HTTP.##35#36{HTTP.http,SubString{String}})() at task.jl:335
ERROR: MethodError: Cannot `convert` an object of type Base.DNSError to an object of type HTTP.Connection
This may have arisen from a call to the constructor HTTP.Connection(...),
since type constructors fall back to convert methods.
Stacktrace:
 [1] HTTP.Connection(::Base.DNSError) at sysimg.jl:24
 [2] getconn(::Type{HTTP.http}, ::HTTP.Client{Base.TTY}, ::String, ::HTTP.RequestOptions, ::Bool) at /Users/jameson/.julia/v0.6/HTTP/src/client.jl:153
 [3] #request#32(::Array{HTTP.Response,1}, ::Bool, ::Bool, ::Function, ::HTTP.Client{Base.TTY}, ::HTTP.Request, ::HTTP.RequestOptions) at /Users/jameson/.julia/v0.6/HTTP/src/client.jl:109
 [4] (::HTTP.#kw##request)(::Array{Any,1}, ::HTTP.#request, ::HTTP.Client{Base.TTY}, ::HTTP.Request, ::HTTP.RequestOptions) at <missing>:0
 [5] #request#30(::Dict{String,String}, ::HTTP.FIFOBuffer, ::Bool, ::Bool, ::Array{Any,1}, ::Function, ::HTTP.Client{Base.TTY}, ::HTTP.Method, ::HTTP.URI) at /Users/jameson/.julia/v0.6/HTTP/src/client.jl:93
 [6] (::HTTP.#kw##request)(::Array{Any,1}, ::HTTP.#request, ::HTTP.Client{Base.TTY}, ::HTTP.Method, ::HTTP.URI) at <missing>:0
 [7] #get#43(::Bool, ::String, ::Array{Any,1}, ::Function, ::String) at /Users/jameson/.julia/v0.6/HTTP/src/client.jl:388
 [8] macro expansion; at REPL[5]:2 [inlined]
 [9] anonymous at <missing>:0
 [10] eval(::Module, ::Any) at ./boot.jl:235
 [11] eval_user_input(::Any, ::Base.REPL.REPLBackend) at REPL.jl:66
 [12] macro expansion; at REPL.jl:97 [inlined]
 [13] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at event.jl:73

```

Even after `gc()`, the count of open descriptors does not decrease.

**Comment by quinnj** [comment]
Should be mostly fixed via 89151e4b639a9fcbad353be621c48b538caf6093 (committed a few hours ago); something got messed up where we weren't re-using connections properly. There is still no hardcap on total # of connections, but they should be properly re-used (to the same host) and cleaned up when the connection is closed now.

**Comment by quinnj** [comment]
Alright, I think with e7bafe3d813ab118f79d038bb422db5bd3520d66 we're good to go here. The previous commit helped, but still didn't completely solve the same symptoms I was seeing in the case of many asynchronous requests. Do re-open if you see any other issues here.
